File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/KrEditModeToolbarUIExtension.cs
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Client.Tiles;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Default.Shared.Workflow.KrPermissions;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Tiles;

namespace Tessa.Extensions.Default.Client.UI.KrProcess
{
    public sealed class KrEditModeToolbarUIExtension : CardUIExtension
    {
        #region Command Actions

        private static async void OpenForEditingAsync(object parameter)
        {
            await KrTileHelper.OpenMarkedCardAsync(KrPermissionsHelper.CalculatePermissionsMark,
                null, //Не требуем подтверждения действия, если не было изменений
                () => TessaDialog.ConfirmWithCancel("$KrTiles_EditModeConfirmation"));
        }

        private static bool CanOpenForEditing(
            object parameter) => UIContext.Current?.CardEditor?.CardModel?.TileIsVisible(DefaultTileNames.KrEditMode) == true;

        private static bool CanOpenForEditing(
            Card card) => card.TileIsVisible(DefaultTileNames.KrEditMode);


        #endregion

        #region Base Overrides

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            var actions = context.ToolbarActions;
            var editModeItem = actions.TryGet(DefaultTileNames.KrEditMode);
            var containsEditMode = editModeItem != null;
            if (!containsEditMode && CanOpenForEditing(context.Card))
            {
                var item = new CardToolbarAction(
                    DefaultTileNames.KrEditMode,
                    "$KrTiles_EditMode",
                    context.UIContext.CardEditor.Toolbar.CreateIcon("Thin2"),
                    new DelegateCommand(OpenForEditingAsync, CanOpenForEditing),
                    tooltip: TileHelper.GetToolTip("$KrTiles_EditModeTooltip", KrTileKeys.Edit),
                    order: 11,
                    gestures: new [] { KrTileKeys.Edit }
                );
                actions.Add(item);
            }
            if (containsEditMode && !CanOpenForEditing(context.Card))
            {
                actions.Remove(editModeItem);
            }
        }

        #endregion
    }
}

File: WebClient SDK/src/default/ui/krProcess/krEditModeToolbarUIExtension.ts
import {
  CardUIExtension,
  ICardUIExtensionContext,
  CardDefaultDialog,
  CardToolbarAction,
  tileIsVisible
} from 'tessa/ui/cards';
import { showConfirmWithCancel } from 'tessa/ui';
import { openMarkedCard } from 'tessa/ui/tiles';

export class KrEditModeToolbarUIExtension extends CardUIExtension {
  public initialized(context: ICardUIExtensionContext) {
    if (context.dialogName !== CardDefaultDialog) {
      return;
    }

    if (tileIsVisible(context.card, 'KrEditMode')) {
      context.toolbar.addItemIfNotExists(
        new CardToolbarAction({
          name: 'KrEditMode',
          caption: '$KrTiles_EditMode',
          icon: 'ta icon-thin-002',
          order: 11,
          toolTip: '$KrTiles_EditModeTooltip',
          command: this.openForEditing
        }),
        {
          name: 'Alt+E',
          key: 'KeyE',
          modifiers: { alt: true }
        }
      );
    } else {
      context.toolbar.removeItemIfExists('KrEditMode');
    }
  }

  private openForEditing = async () => {
    await openMarkedCard(
      'kr_calculate_permissions',
      null, // Не требуем подтверждения действия, если не было изменений
      () => showConfirmWithCancel('$KrTiles_EditModeConfirmation')
    );
  };
}
